embassy-stm32/src/timer/simple_pwm.rs

```diff
@@ -135,13 +152,100 @@ impl<'d, T: CaptureCompare16bitInstance> SimplePwm<'d, T> {
         self.inner.set_compare_value(channel, duty)
     }
 
+    /// Get the duty for a given channel.
+    ///
+    /// The value ranges from 0 for 0% duty, to [`get_max_duty`](Self::get_max_duty) for 100% duty, both included.
+    pub fn get_duty(&self, channel: Channel) -> u16 {
+        self.inner.get_compare_value(channel)
+    }
+
     /// Set the output polarity for a given channel.
     pub fn set_polarity(&mut self, channel: Channel, polarity: OutputPolarity) {
         self.inner.set_output_polarity(channel, polarity);
     }
 }
 
-impl<'d, T: CaptureCompare16bitInstance> embedded_hal_02::Pwm for SimplePwm<'d, T> {
+impl<'d, T: CaptureCompare16bitInstance + Basic16bitInstance, Dma> SimplePwm<'d, T, Dma>
+where
+    Dma: super::UpDma<T>,
+{
+    /// Generate a sequence of PWM waveform
+    pub async fn gen_waveform(&mut self, channel: Channel, duty: &[u16]) {
+        assert!(duty.iter().all(|v| *v <= self.get_max_duty()));
+
+        #[cfg_attr(any(stm32f334, stm32f378), allow(clippy::let_unit_value))]
+        let req = self.dma.request();
+
+        #[cfg(not(any(bdma, gpdma)))]
```

Comment:
we should avoid touching DMA registers from outside the `embassy_stm32::dma` code. Could you move the fifo issue handling into DMA code? Maybe always clear it, maybe add an option.

Also, are you sure it's reuqired to clear FEIF? I'd say it's fine to leave it set forever, it doesn't cause DMA to stop working (or does it?)

Reply:
According to my test, FEIF doesn't affect DMA. I just thought we can take response to clear it.

I just remove all calls to dma regs, leave a note about UDE and FEIF. Once `dma::Transfer::clear_irqs()` or something similar get publiced, we can use it clean the flag.